### PR TITLE
[swift-4.0-branch] Add workaround for MSVC bug

### PR DESCRIPTION
**Pull request by hughbe**
See #33

The problem is that MSVC doesn't recognise the following syntax, reporting multiple errors [here](https://github.com/apple/swift/blob/master/include/swift/AST/Expr.h#L661):
```
  size_t numTrailingObjects(
      typename TrailingObjects::template OverloadToken<SourceLoc>) const {
    return asDerived().hasArgumentLabelLocs()
             ? asDerived().getNumArguments()
             : 0;
  }
```

For some reason, MSVC ignores the fact that we gave friend access to `llvm::TrailingObjects` [here]().

Note that the Swift source code also has to be modified, as this is not enough. This is the new source code, that fully compiles after this change!
```
  // Work around MSVC bug: can't infer llvm::trailing_objects_internal, even though we granted friend access to it
	size_t numTrailingObjects(
		llvm::trailing_objects_internal::TrailingObjectsBase::OverloadToken<Identifier>) const {
		typename TrailingObjects::template OverloadToken<Identifier>) const {
    return asDerived().getNumArguments();
  }
```

### CODE REVIEW

**Comment by jrose-apple** [comment]
Is this already in upstream-with-swift?

**Review by evieluvsrainbows — APPROVED** [review]

**Comment by hughbe** [comment]
Nope: https://reviews.llvm.org/D29880 has been accepted but not yet been merged in.

**Comment by jrose-apple** [comment]
Let's take the upstream fix instead once you get it merged. (If you don't have commit access there you can ask James to merge it for you.)

**Pull request closed without merging**
